src/algebra/monoid_algebra/basic.lean

```diff
@@ -542,14 +542,32 @@ lemma single_one_comm [comm_semiring k] [mul_one_class G] (r : k) (f : monoid_al
 by { ext, rw [single_one_mul_apply, mul_single_one_apply, mul_comm] }
 
 /-- `finsupp.single 1` as a `ring_hom` -/
-@[simps] def single_one_ring_hom [semiring k] [monoid G] : k →+* monoid_algebra k G :=
+@[simps] def single_one_ring_hom [semiring k] [mul_one_class G] : k →+* monoid_algebra k G :=
 { map_one' := rfl,
   map_mul' := λ x y, by rw [single_add_hom, single_mul_single, one_mul],
   ..finsupp.single_add_hom 1}
 
+/--  A monoid homomorphism `f : G →* H` extends to a monoid homomorphism `G →* monoid_algebra k H`,
+mapping `n : G` to `finsupp.single (f x) 1`. -/
+def single_one_right_hom (k : Type*) {H : Type*} [semiring k] [mul_one_class G] [mul_one_class H]
+  (f : G →* H) : G →* monoid_algebra k H :=
+(of k H).comp f
+
+@[simp] lemma single_one_right_hom_apply (k : Type*) {H : Type*} [semiring k] [mul_one_class G]
+  [mul_one_class H] (f : G →* H) (n : G) :
+  single_one_right_hom k f n = single (f n) 1 := rfl
+
+/--  A multiplicative homomorphism `f : G →* H` between two monoids induces a `k`-algebra
+homomorphism `monoid_ring_hom_map f : monoid_algebra k G →+* monoid_algebra k H`. -/
+def monoid_ring_hom_map (k : Type*) {H : Type*} [semiring k] [monoid G] [monoid H]
+  (f : G →* H) : monoid_algebra k G →+* monoid_algebra k H :=
+lift_nc_ring_hom single_one_ring_hom (single_one_right_hom k f) $ λ r n,
+show (single 1 r * (single (f n) 1) : monoid_algebra k H) = single (f n) 1 * (single 1 r),
+by simp only [single_mul_single, mul_one, one_mul]
```

Comment:
This implementation also almost works:
```lean
{ map_one' := sorry,
  map_mul' := λ x y, @map_domain_mul _ _ _ _ _ _ x y f.to_mul_hom,
  ..(finsupp.map_domain.add_monoid_hom f :  monoid_algebra k G →+ monoid_algebra k H) }
```

Reply:
I am probably confused, but it seems that `ma_domain` is going to assign a value, by summing over all elements in the fiber, right?  If so, this is not the map that I want.

Reply:
#13389 proves the `sorry` and removes the need for the `@`

Reply:
Eric, thank you for your help!  I will wait for the other PR to get merged: I have been a little stumped by getting everything to work for a `semi_ring` and I hope that with your new additions it will work smoothly.

Reply:
I'm happy with your definition of the map, but I'm having a hard time using it.

In particular, proving that a monomial in `R[X]` converts to a `finsupp.single : R [T;T⁻¹]` (in the context of Laurent polynomials), is eluding me.  I boiled it down to showing R-linearity and the fact that `X ↦T`.

I think that there is a single `rfl`-lemma showing how to apply your map that will fix things, but I've been unable to find it.

Do you have any suggestions?

Thanks!

Reply:
Aren't you looking for `map_domain_single`?

Reply:
Turns out I was, thanks!  I have to `convert` it, but it works!